lib/WUI/nhttp/segmented_json.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <cstdlib>
+#include <cstdint>
+#include <tuple>
+
+namespace nhttp {
+
+/// Support for rendering JSON by parts.
+///
+/// Sometimes, we may not afford to send the whole JSON answer in one packet
+/// and we have only a single packet of buffer space. Therefore, we need to be
+/// able to store just enough of state to be able to generate a consistent
+/// reply and then allow suspending at places.
+///
+/// This supports suspending between individual JSON fields. We still assume
+/// we'll be able to put at least the field inside the buffer (that is, no extra
+/// huge strings).
+///
+/// The user inherits this, adds the state and provides the content function.
+///
+/// This does not validate the produced JSON in any way (and probably could be
+/// used to render something else than JSON, actually).
+///
+/// Few notes for implementers:
+/// * The idea behind the content function is that it tries to call methods on
+///   the provided output until it returns ContentResult::Incomplete. Then the
+///   content function returns and it'll get called in the future with a new buffer
+///   and the resume_point parameter set to whatever was passed to the output that
```

Comment:
Do I understand correctly, that you render the JSON as many times as many segments are used? I.e. when rendering the second segment, you just skip dumping the first characters until you reach the resume_point offset?

Reply:
No. The internal (hidden through the macros) switch is used as kind of a goto and it jumps to the place that didn't fit previously. So that one is retried, but all the old ones are just skipped and not even called.